Content.Shared/Weapons/Ranged/Systems/BatteryWeaponFireModesSystem.cs

```diff
@@ -74,12 +80,12 @@ private void OnGetVerb(EntityUid uid, BatteryWeaponFireModesComponent component,
         }
     }
 
-    private void OnInteractHandEvent(EntityUid uid, BatteryWeaponFireModesComponent component, ActivateInWorldEvent args)
+    private void OnUseInHandEvent(EntityUid uid, BatteryWeaponFireModesComponent component, UseInHandEvent args)
     {
-        if (!args.Complex)
+        if (component.FireModes.Count < 2)
             return;
 
-        if (component.FireModes.Count < 2)
+        if (TryComp<AccessReaderComponent>(uid, out var accessReader) && !_accessReaderSystem.IsAllowed(args.User, uid, accessReader))
             return;
```

Comment:
This should be in cycle or a trycycle method rather than every caller having to handle it.

Reply:
Thinking about it I would just split this out to a new component, make an event like "CycleAttemptEvent", and then also make the other cycle code use it.